rfcs/convergence/simplify-prop-merging.md

```diff
@@ -0,0 +1,359 @@
+
+```tsx
+// ComponentState
+const state = {
+  buttonSlot: { as: 'button', type: 'submit' },
+};
+
+function App() {
+  // Typescript cannot warn that this is potentially evil
+  return <Component buttonSlot={{ as: 'div' }} />;
+}
+```
+
+```html
+<!-- Thankfully still works, but props must be filtered on every render -->
+<div></div>
+```
+
+The above example still works thanks to the the native attribute filtering that is done as part of `getSlots`. However, these filters always need to be run when resolving shorthand slots. During each render, every prop of a component is iterated and re-filtered.
+
+The current filtering also needs to iterate all the allowed props during a nested loop, although this could be optimized quite easily.
+
+While the existence of these helpers is necessary (as explained in the below proposals), it seems unnecessary that it needs to happen on every render.
+
+### Extending a component by adding/modifying slots
+
+One common use case as seen in `Menu`, `Avatar` and `Badge` is to extend a base component by adding or modifying a slot. There are two ways where it is possible to do this currently:
+
+```tsx
+// Extend by modifying existing slots
+// call resolveShorthandprops on defaultProps
+const state = mergeProps(
+  {
+    as: 'span',
+    label: { as: 'span' },
+    getInitials: defaultGetInitials,
+    ref,
+  },
+  defaultProps && resolveShorthandProps(defaultProps, avatarShorthandProps),
+  resolveShorthandProps(props, avatarShorthandProps),
+);
+```
+
+```tsx
+// Extend by adding a slot
+// Mutate state in extended component and call mergeProps again
+const baseState = useBaseComponent(props, ref);
+
+// React elements cannot be extended and will break `resolveShorthandProps`
+// set to undefined since it will be resolved again anyway
+(baseState as unknown) as ExtendedComponentProps).slot = undefined;
+const state = mergeProps(baseState, defaultProps, resolveShorthandProps(props, extendedComponentShorthandProps));
+```
+
+Both of the examples above are quite opinionated and require knowledge of the internals of `mergeProps` and `resolveShorthands` to understand without guidance.
+
+Ideally, conformance tests would be required to validate `defaultProps` always goes through `resolveShorthandProps`.
+
+In the case where a component is extended by adding slots, it's pretty hard to validate both in terms of functionality and type safety.
+
+### Merging functions and properties with primitives
+
+Functions or primitives used in `mergeProps` will simply replace the value in the order that they were defined. This is extremely similar to native object spreading but without a deep merge functionality.
+
+Prop objects tend to be relatively flat, and in cases where objects/arrays are used that need to be deep merged on each render we should examine those cases and make exceptions.
```

Comment:
can we move the example that is provided at the beginning of the RFC here, so we have unified format and examples provided within chapters ?